### ISSUE TITLE
Does `optimizer.step()` function complete gradient update before next batch coming in?

### ISSUE THREAD

**Issue by zarzen** [horovod/horovod#1491]
**Environment:**
1. Framework: PyTorch 
2. Framework version: 1.3.0
3. Horovod version: 0.18.2
4. MPI version: 4.0.2
5. CUDA version: X
6. NCCL version: X
7. Python version: 3.7.1
8. OS and version: Ubuntu 16.04
9. GCC version:

**Your question:**
Does `horovod.torch.DistributedOptimizer.step()` function complete the gradient updates before next batch started forwarding process?
I have timed the `optimizer.step()` function. It only takes around 100ms to complete, under 1Gbps link for resnet50 model. The gradient size is around 100MBytes, which is unreasonable to have the `step()` function complete in such a short time (It should take around 800ms to complete sync). In this case, when the next `forwarding` process starting, the gradients of the model won't be possibly updated. So is this true that model won't wait for gradients to be fully updated when doing the forwarding process?

**Comment by tgaddair** [comment]
Hey @zarzen, the allreduce will actually occur after the call to `loss.backward()` via a hook.  When `optimizer.step()` is called, the optimizer will synchronize to ensure all gradients have been aggregated.  So you would probably want to time both methods together.

**Comment by zarzen** [comment]
Thanks!
